src/serac/physics/utilities/functional/functional.hpp

```diff
@@ -292,6 +292,16 @@ class Functional<test(trial)> : public mfem::Operator {
     Evaluation<Operation::GradientMult>(input_T, output_T);
   }
 
+  /**
+   * @brief Obtains the element stiffness matrix reshaped a mfem::Vector
+   * @param[inout] K_e A mfem::Vector containing the assembled element stiffness matrix (test_dim * test_ndof, trial_dim
+   * * trial_ndof, nelem)
+   */
+  virtual void GradientMatrix(mfem::Vector& K_e) const
+  {
+    for (auto domain : domain_integrals_) domain.GradientMatrix(K_e);
+  }
```

Comment:
Does this need to be `virtual`? 

The function is also a little bit misleading, since it is returning a packed array of element matrices, not the actual gradient matrix itself. Maybe `ComputeElementMatrices()` or something similar would be clearer (or `GradientMatrix` could return the `AssembledSparseMatrix` object?)

I prefer to return values to out parameters when possible, but I'm not sure if we have anything about that choice in our style guide.

Reply:
Great suggestions and questions. As is, I don't think it needs to be `virtual`. I think it depends on how we want it to interface with CUSPARSE?

I intended the sparsity pattern to be encoded when constructed and updated as we go along. Since every new state will involve new values, but not necessarily new sparsity, I didn't want to return another `AssembledSparseMatrix`? I also wanted to be able to accumulate various domain_integrals as we go along. I think ideally on the GPU we want a block of memory that's pre-allocated that we just change the values directly on the GPU itself.

Reply:
Maybe one possibility is to pass in an `AssembledSparseMatrix`. We can then do some check on that and encode it directly into the final storage pattern. While `K_e` is a temporary, it's matrix-format agnostic; I know in the past we had talked about using a non-conventional CSR format also.

Reply:
Another option is to make the `Functional` object maintain a `AssembledSparseMatrix` as a member variable. If the user calls `GradientMatrix`, it could form that matrix (and only figure out sparsity on the first call to `GradientMatrix`, caching it on subsequent calls), and return a reference to it.

Reply:
@samuelpmishLLNL I think that might work nicely. I'd like to keep support for `K_e` though, as it would make it easier to compare new matrix formats and so forth. Once you "assemble", it will be difficult to retrieve `K_e`.

Reply:
There are definitely cases where you want just the element matrices, but I would like to also provide an interface that doesn't require the user jumping through too many hoops to just get the gradient matrix. Something like this would allow for both use cases:

```cpp
class Functional{
  AssembledSparseMatrix K; // gradient matrix
  mfem::Vector K_e; // element stiffness matrices packed into 1D array

  ...

  const auto & ElementMatrices(){
    calculateElementMatrices(K_e);
    return K_e; 
  }

  const auto & GradientMatrix(){
    calculateElementMatrices(K_e);
    return K = assemble(K_e);
  }
};

Reply:
I changed it now to build an AssembledSparseMatrix on the first call to "GradientMatrix()". It then always returns the internal K.